### ISSUE TITLE
[FCM][NS+NG][Android] Handle notification data when app started from notification

### ISSUE THREAD

**Issue by echap** [EddyVerbruggen/nativescript-plugin-firebase#297]
Hi there,

For more than a week now I'm struggling with fcm notification. When the application is loaded from a tap on a notification containing also data.

My goal is to **reroute user on a specific component when I detect that the app started from a notification click**

As of today, I retrive the informations from the app using :

my App.component.ts
```
application.on(application.resumeEvent, this.zoneWrap((args: application.ApplicationEventData) => {
            if (args.android) {
            let intent = args.android.getIntent();
            if (this.identifyNotificationCallback(intent)) {
                     //We clicked on a notification
                try {
                    let params = JSON.parse(intent.getExtras().getString("internalURL")); //Provided by FCM, as a child of data provided by my backend
                    this.routerExtensions.navigate([internalURL]);
                } catch (error) {
                    console.log(error);
                }
            }
        }
}));

```

But the routing event appends not at the good moment, something is going really wrong and I don't know how to handle it.

Help is really much welcomed on this !

**Comment by jorgelcf** [comment]
Hi echap. 

I was having a weird behavior in my Android NG App when I tap on a notification in the status bar similar to what you describe.  The tap event first fires a launch event and then a resume event causing my app.component to 'restart' when in the background. I finally get it to work the same as when the app is resumed by setting **android:launchMode="singleTop"** in the com.tns.NativeScriptActivity in the Manifest.  But, now I am struggling to get the notification intent (extras) since the onNewIntent() method of the activity is called instead of the onCreate(). Reference:[https://developer.android.com/guide/topics/manifest/activity-element.html#lmode]

**Comment by echap** [comment]
@jorgelcf Yes that's actually the kind of behaviors that I have. I belive that this behavior is generally not expected by nativescript angular.

I will try to use your method that may resolve the issue, but I'm afraid of the potential side effects that it can have. Have you noticed some ?

@EddyVerbruggen What have you any clue on this ?

**Comment by jorgelcf** [comment]
@echap I have not noticed any side effect in my tests.

**Comment by kofrimpong** [comment]
@echap  have you found any solution to the singletop launchmode causing the callback not be called?

**Comment by jorgelcf** [comment]
@kofrimpong I got it working this way:
1. Extend the android Activity by creating a copy (just to have the same as what is shipped) of the activity file shipped with NS core modules and add this:
 **NativeScriptActivity.prototype.onNewIntent = function (intent) {
        console.info('onNewIntent', intent);
        _super.prototype.onNewIntent.call(this, intent);
        this.setIntent(intent);
    };**
2. Add this to the app.component.ts (if using Angular):
 **Application.android.on(Application.AndroidApplication.activityResumedEvent, (args: 
     Application.AndroidActivityBundleEventData) => {
            if (args.activity.getIntent().getExtras()) {
               // do your stuff here!
	    }
  });**
3. FYI.. I am using singleTask for the launchmode. If you want to navigate to a page based on the message received data and you face problems, try wrapping the action in a ngZone.

**Comment by simonettoa** [comment]
Hi @jorgelcf ,
I'm having trouble to solve this issue in android (ios works fine). I set the singleTask as suggested from point 3. I set the point two correctly but I have no idea how and where to write the point one you wrote?!?!

If I send a notification (from the terminal using CURL command) and the app is in foreground it takes care of it from the 'addOnMessageReceivedCallback' function, in background the notification arrives but when I click on it I don't have any Extras from the getExtras function in the resumeEvent.
Here what I've got from the (args.android).getIntent() function:

`intent value -> Intent { act=android.intent.action.MAIN cat=[android.intent.category.LAUNCHER] flg=0x10200000 cmp=org.nativescript.TruckMe/com.tns.NativeScriptActivity }`

Any help is much appreciated!

**Comment by pap5508** [comment]
@simonettoa : how you handle push notification in iOS please share some code to handle it ?

**Comment by hettiger** [comment]
Probably related to #692 

@EddyVerbruggen 
I don't understand why this issue got closed. This is obviously a problem with FCM and Dynamic Links. There should be a clean solution. (E. g. callbacks should fire like they are supposed to and the docs should tell one to use another launchMode.) I have also provided some more info on the issue #692 But I'm not really sure what to tell the {NS} team. They probably think it's this plugins responsibility, wont they? This is currently the only thing that's holding me back from upgrading to {NS} 4.0.
